### PR TITLE
REST API to Fetch List of Algorithms

### PR DESCRIPTION
**Pull request by terrymyc**
This PR develops a REST API to fetch a list of available classifiers, balance strategies, query strategies, and feature extraction methods in the backend and implements them in the frontend.
This PR also develops a REST API to clear the error state file while retrying setup.
This PR also adds labels to all algorithms.

image

image

image

### CODE REVIEW

**Comment by J535D165** [comment]
Nice! Might be interesting to consider updating this at the same time https://github.com/asreview/asreview/blob/a296205f7fe1f2e4a09e1856bd5be01438d9223a/asreview/entry_points/algorithms.py#L37-L72. Labels can be interesting to have there as well (`asreview algorithms`)

**Comment by terrymyc** [comment]
> . Labels can be interesting to have there as well (`asreview algorithms`)

@J535D165 So like replace "logistic" with "Logistic regressions", right?

**Comment by J535D165** [comment]
I think both (such that you know the name for the simulation CLI).

**Comment by terrymyc** [comment]
@J535D165 There is an internal bug when model training fails. For example, if I choose Doc2Vec without `gensim` package installed, the model training will throw `ImportError` with `result.json` and `error.json` created. The problem is that if I refresh and choose another viable option other than `Doc2Vec`, the `ImportError` of `gensim` still appears. The moment I change the option, `kwargs.json` will change, but `result.json` and `error.json` are not cleared.

**Comment by J535D165** [comment]
Very interesting feature. Do we need to update parts of the documentation as well?

**Comment by terrymyc** [comment]
> Very interesting feature. Do we need to update parts of the documentation as well?

@J535D165 I think `list_classifiers` and other similar functions need new descriptions. I guess in the documentation those are updated automatically?

File: asreview/models/classifiers/lstm_pool.py [hunk 1]
```diff
@@ -90,6 +90,7 @@ class LSTMPoolClassifier(BaseTrainClassifier):
     """
 
     name = "lstm-pool"
+    label = "LSTM-pool"
```
Comment (J535D165) [inline asreview/models/classifiers/lstm_pool.py:93]: LSTM with max-pooling layer?

Comment (terrymyc) [reply asreview/models/classifiers/lstm_pool.py:93]: Meaning...?

Comment (terrymyc) [reply asreview/models/classifiers/lstm_pool.py:93]: Then what about the label for `lstm-base`?

File: asreview/models/query/max_random.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    At each query 95% of the instances would be sampled with the maximum
+    query strategy after which the remaining 5% would be sampled with
+    the random query strategy.
+    """
+
+    name = "max_random"
+    label = "Mixed (95% Maximum and 5% Random)"
+
+    def _query(self, X):
+        return MixedQuery
```
Comment (terrymyc) [inline asreview/models/query/max_random.py:33]: @J535D165 This subclass seems to work. Did I miss something?

**Comment by J535D165** [comment]
For readers of the https://www.nature.com/articles/s42256-020-00287-7 publication. This PR breaks the naming of "Mixed sampling". 

> The available query strategies for the active learning part are (1) random selection, ignoring model-assigned probabilities; (2) uncertainty-based sampling, which chooses the most uncertain record according to the model (that is, closest to 0.5 probability); (3) certainty-based sampling (max in ASReview), which chooses the record most likely to be included according to the model; and (4) mixed sampling, which uses a combination of random and certainty-based sampling.

Mixed sampling is a contraction of two sampling methods. This is made explicit from now on and replaced by `max_uncertainty` or `max_random`. We decided to not make this partly backward compatible. Users of ASReview LAB/BUI can continue with projects making use of this method. For sake of maintainability, the simulation CLI doesn't have this option anymore. If there are questions about the deprecation, please leave a message in this thread.

File: asreview/models/balance/double.py [hunk 3]
```diff
@@ -49,6 +49,7 @@ class DoubleBalance(BaseBalance):
     """
 
     name = "double"
+    label = "Double"
```
Comment (Rensvandeschoot) [inline asreview/models/balance/double.py:52]: Can you add a bit more explanation like you do with the other labels? for example `Two way rebalancing function`

File: asreview/models/balance/triple.py [hunk 4]
```diff
@@ -62,6 +62,7 @@ class TripleBalance(BaseBalance):
     """
 
     name = "triple"
+    label = "Triple"
```
Comment (Rensvandeschoot) [inline asreview/models/balance/triple.py:65]: Can a bit more information be inlcluded? `This divides the training data into three sets: included papers, excluded    papers found with random sampling and papers found with max sampling`

File: asreview/models/balance/undersample.py [hunk 5]
```diff
@@ -34,6 +34,7 @@ class UndersampleBalance(BaseBalance):
     """
 
     name = "undersample"
+    label = "Undersampling"
```
Comment (Rensvandeschoot) [inline asreview/models/balance/undersample.py:37]: For the label description: `undersampling the irrelevant records`

File: asreview/models/classifiers/lstm_base.py [hunk 6]
```diff
@@ -86,6 +86,7 @@ class LSTMBaseClassifier(BaseTrainClassifier):
     """
 
     name = "lstm-base"
+    label = "LSTM classic"
```
Comment (Rensvandeschoot) [inline asreview/models/classifiers/lstm_base.py:89]: what is the difference between `base` and `classic`? If there is none, I would not suggest to use both terms. Maybe a more informative label can be provided? `LSTM layer with one output, a dense layer, and a single sigmoid output node`

File: asreview/models/feature_extraction/doc2vec.py [hunk 7]
```diff
@@ -90,6 +90,7 @@ class Doc2Vec(BaseFeatureExtraction):
     """
 
     name = "doc2vec"
+    label = "Doc2Vec"
```
Comment (Rensvandeschoot) [inline asreview/models/feature_extraction/doc2vec.py:93]: maybe change to something like `Feature extraction technique provided by the `gensim    &lt;https://radimrehurek.com/gensim/&gt;`__ package`

File: asreview/models/feature_extraction/embedding_idf.py [hunk 8]
```diff
@@ -63,6 +63,7 @@ class EmbeddingIdf(BaseFeatureExtraction):
     """
 
     name = "embedding-idf"
+    label = "Embedding IDF"
```
Comment (Rensvandeschoot) [inline asreview/models/feature_extraction/embedding_idf.py:66]: maybe add a bit more info?

File: asreview/models/feature_extraction/embedding_lstm.py [hunk 9]
```diff
@@ -75,6 +75,7 @@ class EmbeddingLSTM(BaseFeatureExtraction):
     """
 
     name = "embedding-lstm"
+    label = "Embedding LSTM"
```
Comment (Rensvandeschoot) [inline asreview/models/feature_extraction/embedding_lstm.py:78]: same.. needs to be a bit more informative

File: asreview/models/query/max.py [hunk 10]
```diff
@@ -19,12 +19,13 @@
...
 
 class MaxQuery(ProbaQueryStrategy):
-    """Maximum sampling query strategy.
+    """Maximum query strategy.
 
     Choose the most likely samples to be included according to the model.
     """
 
     name = "max"
+    label = "Maximum"
```
Comment (Rensvandeschoot) [inline asreview/models/query/max.py:28]: `Max refers to the most likely records to be shown to classify`

File: asreview/models/query/max.py [hunk 11]
```diff
@@ -19,12 +19,13 @@
 
 
 class MaxQuery(ProbaQueryStrategy):
-    """Maximum sampling query strategy.
+    """Maximum query strategy.
 
     Choose the most likely samples to be included according to the model.
```
Comment (Rensvandeschoot) [inline asreview/models/query/max.py:24]: samples=>records?
included=>shown?

File: asreview/models/query/random.py [hunk 12]
```diff
@@ -20,7 +20,7 @@
 
 
 class RandomQuery(NotProbaQueryStrategy):
-    """Random sampling query strategy.
+    """Random query strategy.
 
     Randomly select samples with no regard to model assigned probabilities.
```
Comment (Rensvandeschoot) [inline asreview/models/query/random.py:25]: samples=>records?

File: asreview/models/query/random.py [hunk 13]
```diff
@@ -30,6 +30,7 @@ class RandomQuery(NotProbaQueryStrategy):
 
     """
     name = "random"
+    label = "Random"
```
Comment (Rensvandeschoot) [inline asreview/models/query/random.py:33]: `Random selection of records`?

File: asreview/models/query/uncertainty.py [hunk 14]
```diff
@@ -19,7 +19,7 @@
 
 
 class UncertaintyQuery(ProbaQueryStrategy):
-    """Maximum uncertainty query strategy.
+    """Uncertainty query strategy.
 
     Choose the most uncertain samples according to the model (i.e. closest to
```
Comment (Rensvandeschoot) [inline asreview/models/query/uncertainty.py:24]: samples=>records

File: asreview/models/query/uncertainty.py [hunk 15]
```diff
@@ -28,6 +28,7 @@ class UncertaintyQuery(ProbaQueryStrategy):
     """
 
     name = "uncertainty"
+    label = "Uncertainty"
```
Comment (Rensvandeschoot) [inline asreview/models/query/uncertainty.py:31]: `Shows the most uncertain records to be screened`

File: asreview/models/query/uncertainty.py (hunk 14 above)
Comment (J535D165) [reply asreview/models/query/uncertainty.py:24]: Out of the scope of this PR.

File: asreview/models/query/random.py (hunk 12 above)
Comment (J535D165) [reply asreview/models/query/random.py:25]: Out of the scope of this PR.

File: asreview/models/query/max.py (hunk 11 above)
Comment (J535D165) [reply asreview/models/query/max.py:24]: Outside the scope of this PR.

**Comment by J535D165** [comment]
@Rensvandeschoot this many changes compared to https://www.nature.com/articles/s42256-020-00287-7/tables/2 will be very confusing and requires major issues with our documentation, isn't it?

I would suggest replacing the labels of double and triple with Dynamic resampling (Double) and Dynamic resampling (Triple) (see https://www.nature.com/articles/s42256-020-00287-7/tables/2). I would suggest not making major changes to the other labels.

File: asreview/models/balance/double.py (hunk 3 above)
Comment (J535D165) [inline asreview/models/balance/double.py:52]:
```suggestion
    label = "Dynamic resampling (Double)"
```

File: asreview/models/balance/triple.py (hunk 4 above)
Comment (J535D165) [inline asreview/models/balance/triple.py:65]:
```suggestion
    label = "Dynamic resampling (Triple)"
```

**Comment by J535D165** [comment]
YESSSS. Thank you so much @terrymyc!

**Pull request merged**